### PR TITLE
reimplemented "ban bad peers" feature

### PR DESCRIPTION
**Pull request by Kolcha**
first of all, I liked features provided by this qBittorrent variant, but I think their implementation is far to be optimal. this pull request suggest much more simpler (and cleaner) implementation of peer filtering feature (that's why I found that project).

firstly, few words about currently existing logic. some timer periodically gets list of all connected peers from all torrents and then bans some of them depending on settings and rules. peer ban is achieved by adding "bad peer" IP to libtorrent peer blacklist. this is very inconvenient and expensive: get (and copy) list of already banned IPs known to libtorrent, add new IP (this can cause memory re-allocate operation), and then return modified list to libtorrent (very likely another copy is happen). moreover, iterating over thousands of torrents (and likely peers) on timer may have great performance impact and this is not guarantee that nothing is downloaded by bad peer. this is from performance point of view. from logic point of view, it is very unclear to me why some "unban" timer exists, which unbans peers previously considered "bad" and banned. also from implementation point of view, there are a lot of complex filter conditions in one place and very unclear path how peer is banned.

what I suggest: do NOT **ban peer** using libtorrent "blacklist" API: it is inconvenient and pretty useless - because IP may change, also do NOT use timers to analyze ALL connected peer at one time. instead of that, I suggest just **drop connections** from such "bad" peers on early stage, almost after the moment they establish connection and we consider them "bad". this logic doesn't require "blacklist" editing and handles peers as they arrive (i.e. "on the fly"). in that case NOTHING will be uploaded to banned peers at all. in most cases and few (or even one) rejected connection peers will stop connecting to us, considering us as dead/unreachable or similar. moreover, this logic will perfectly handle IP address changes (same client by with different address will dropped). this behavior can be achieved easily using peer and torrent libttorent extensions. also such implementation is very easily extensible and localized in few separated files, which makes changes moves with upstream version releases more easy.

the filters I implemented is completely the same as original, they are just moved to another place and maybe slightly simplified. behavior is the same, nothing was changed. maybe only one thing to notice - **banned peers are not written to log**, they are **written into SQLITE database**. this is because I don't think that the log is good place to show banned peers, and in my case peers are **not banned**, they **are just dropped**, and such event can happen very often, what will only flood the logs, but database will have only one record per dropped peer.

changes were tested by downloading and seeding about a day 2 torrents from #127 , generated database containing list of banned peers can be [downloaded here](https://www.dropbox.com/s/kintsrlwyixvzn1/peers.db?dl=1).

P.S> I have some more ideas how to improve suggested enhanced feature, very likely I'll open issue (or few) for discussion, because some points of existing implementation are unclear to me. but this is later, I have not so much time to spent on hobby.

### CODE REVIEW

**Comment by c0re100** [comment]
Thank you for reimplement!!!
I found out that libtorrent extension `peer_connection_handle` can drop connection in late August(when I'm debugging #127), but I don't have time to reimplement ban peer feature.

`it is very unclear to me why some "unban" timer exists`
Well, 'ban peer' logic written 4 years ago (2016), and I just started learing C++ at this moment (I haven't see qbt codebase when I started writing this feature--yes, it's trial and error....) , sorry for this spaghetti code....
Anyway, as you know, I haven't implement `ban peer` feature with libtorrent extensions.
So I need to unban banned IP after 60mins.
Basically, it works and I don't touch it again LOL.

Yes, I know many enhanced features need to refactoring, e.g.: offline downloader IP parser.

PS: I will test it at Sat

File: src/base/bittorrent/peer_blacklist.hpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  std::string pid = info.pid.to_string().substr(0, 8);
+  std::regex player_filter("-(UW\\w{4})-");
+  return !!std::regex_match(pid, player_filter);
+}
+
+
+// drop connection action
+void drop_connection(lt::peer_connection_handle ph)
+{
+  ph.disconnect(boost::asio::error::connection_refused, lt::operation_t::bittorrent, 0);
```
Comment (c0re100) [inline src/base/bittorrent/peer_blacklist.hpp:48]: Last parameter `disconnect_severity_t` is optional, so you can remove it.
```
void disconnect (error_code const& ec, operation_t op
    , disconnect_severity_t = peer_connection_interface::normal);
```

Comment (Kolcha) [reply src/base/bittorrent/peer_blacklist.hpp:48]: unfortunately not... I even can't use constants... using constant (or dropping it and using default value) leads to linker error:
```
undefined reference to `libtorrent::peer_connection_interface::normal'
```
this happens at least on Debian stable and Ubuntu 20.04, and only when such constants are used in header files, everything is when they are used in .cpp. and this behavior is compile-specific, this issue is not reproducible (code compiles fine) on macOS (where I initially wrote this code) and Debian unstable).
building qBittorrent in C++17 mode also fixes that issue: https://stackoverflow.com/questions/40690260/undefined-reference-error-for-static-constexpr-member
so, that '0' at the end is workaround to guarantee successful build on almost any compiler

Comment (c0re100) [reply src/base/bittorrent/peer_blacklist.hpp:48]: I can't reproduce (`libtorrent::peer_connection_interface::normal'`) on MSVC, but MSVC don't like `0` for workaround, even change to C++17 mode.
```
Error	C2664	'void libtorrent::peer_connection_handle::disconnect(const libtorrent::error_code &,libtorrent::operation_t,libtorrent::disconnect_severity_t)': cannot convert argument 3 from 'int' to 'libtorrent::disconnect_severity_t' (compiling source file E:\qBittorrent\qBittorrent430\src\base\bittorrent\session.cpp)	qbt_base	E:\qBittorrent\qBittorrent430\src\base\bittorrent\peer_blacklist.hpp	48	
```

No problem if using jom for building, weird...

How about use `lt::disconnect_severity_t{0}` for workaround?
It is fine for MSVC and ubuntu 20.04.

Comment (Kolcha) [reply src/base/bittorrent/peer_blacklist.hpp]: done, GCC10 on Debian unstable is happy with it too, thanks!

**Comment by c0re100** [comment]
Thank you!

**Comment by c0re100** [comment]
I discovered this ban bad peers plugin consume lots of CPU resources and client name filter not working too...
high CPU usage (when session have many connections):
photo_2020-11-30_05-54-50
It look like consume too much resource on each `on_interested/on_have........` call.
Just leave handshake message handler, then everything fine, CPU back to around 35-50% like old implementation.
For testing client name filter, you can add BitComet/Deluge or qBittorrent itself to regex.
